File: Solutions/0251-flatten-2d-vector/solution.cpp
class Vector2D {
public:
    vector<vector<int>>* arr;
    int i = 0, j = 0;
    
    Vector2D(vector<vector<int>>& vec) {
        arr = &vec;
    }
    
    int next() {
        goton();
        vector<vector<int>> vec = *arr;
        return vec[i][j++];
    }
    
    bool hasNext() {
        vector<vector<int>> vec = *arr;
        goton();
        return i < vec.size();
    }

    void goton() {
        vector<vector<int>> vec = *arr;
        while(i < vec.size() && j == vec[i].size()) {
            j = 0;
            i++;
        }
    }
};

/**
 * Your Vector2D object will be instantiated and called as such:
 * Vector2D* obj = new Vector2D(vec);
 * int param_1 = obj->next();
 * bool param_2 = obj->hasNext();
 */

File: Solutions/0251-flatten-2d-vector/solution.js
/**
 * @param {number[][]} vec
 */
var Vector2D = function(vec) {
    this.arr = [];
    for(let el of vec) {
        for(let elel of el) {
            this.arr.push(elel);
        }
    }
    this.i = 0;
};

/**
 * @return {number}
 */
Vector2D.prototype.next = function() {
    return this.arr[this.i++];
};

/**
 * @return {boolean}
 */
Vector2D.prototype.hasNext = function() {
    return this.i < this.arr.length;
};

/** 
 * Your Vector2D object will be instantiated and called as such:
 * var obj = new Vector2D(vec)
 * var param_1 = obj.next()
 * var param_2 = obj.hasNext()
 */
